### ISSUE TITLE
Alternator: add CRC header

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5039]
DynamoDB responses have a "“x-amz-crc32” header with a checksum of the response. By inspecting the Java client, I see that the CRC algorithm used is the same one as in java.util.zip.CRC32, which is in turn the same one defined in RFC 1952 for the gzip format. There are some questions on what exactly gets CRCed and how this interacts with compression.

This is a low priority issue, because all clients that we tried don't seem to care that the CRC header is missing.

**Comment by tzach** [comment]
>  because all clients that we tried don't seem to care that the CRC header is missing.

Indeed. This is from AWC CLI in debug mode:
```
2019-09-16 13:11:46,602 - MainThread - botocore.retryhandler - DEBUG - crc32 check skipped, the x-amz-crc32 header is not in the http response.
```

**Comment by nyh** [comment]
@nuivall noticed that the Golang library prints a warning when the response's CRC doesn't match (or in our case, is missing).

**Comment by nuivall** [comment]
More specificaly this library: `./github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.17.1/internal/customizations/checksum.go`, it can flood user stdout / logging system. 

It prints `WARN failed to close HTTP response body, this may affect connection reuse response did not match expected checksum`. But as far as I can tell this particular error won't result in connection reuse for this library.

**Comment by xyu335** [comment]
Is there any updates on this issue to add the same CRC in response header? using `github.com/aws/aws-sdk-go-v2 v1.20.0`, i still have this error logs printed.

**Comment by nyh** [comment]
Although it shouldn't be too hard to implement this CRC header and close this issue, I'm reluctant to implement it because I view it as an anti-feature: it adds more computation to each and every request, for no apparent (to me...) reason. TCP already has CRC-based validation (and obviously, TLS has even better), so why add another one? Is it to protect against corruption some HTTP proxy? Is this a common problem?

I looked at the **Python** SDK (botocore/retryhandler.py), and it clearly distinguishes between two different cases:
1. If the CRC header is missing in the response, a debug-level message is printed "crc32 check skipped, the %s header is not in the http response.". By default debug-level messages aren't printed and nothing happens - the driver behaves as if the checksum check was successful.
2. If the CRC header is present, only then does the SDK calculate the CRC of the response and checks its equality to the header's content. If they don't match, an exception is raised and the request is retried.

If the Go SDK does something different when the CRC header is completely missing, and prints visible (not "debug-level") messages and even closes connections (?), perhaps it should be reported to the Go SDK developers as a bug, citing the Python SDK's behavior is the correct one?

**Comment by nuivall** [comment]
1. There is an option to disable validation in aws-sdk-go-v2:
```
	// Allows you to disable the client's validation of response integrity using CRC32
	// checksum. Enabled by default.
	DisableValidateResponseChecksum bool
```
2. We could try to send a patch to aws-sdk-go-v2 as code looks a bit broken, it sets `ok` variable but never use it, simple addition like `if !ok { /*warn*/ }` would turn error into warning

**Comment by nyh** [comment]
> There are some questions on what exactly gets CRCed and how this interacts with compression.

I wrote a test for this, and the answer is clear-cut: DynamoDB calculates the CRC over the **compressed** response - the response body actually sent over the wire. This choice caused problems for various HTTP libraries which by default automatically decompress the response before ever returning it to the user - because at that point the user is no longer able to check the CRC. This also lead most SDKs to outright ignore these CRCs.

Starting to implement this feature also brings up a frustrating problem: To calculate the CRC and place it in the header, before the body, the implementation is forced to buffer the entire response in memory and not "stream" it in pieces. This suggests that DynamoDB (which always calculates this CRC) always buffers the entire response and doesn't stream it. Another piece of evidence is that experimentally we noticed that DynamoDB **never** uses chunked encoding, not even for large response (the largest can be a BatchGetItem limited to 16 MB). It is beginning to appear that in Alternator, if we want to enable CRC we'll need to disable streaming responses - even for large responses - and chunked encoding, to allow us to calculate the CRC header before starting to send the response body. Additionally, if response compression is active, we will need to compress the entire body into memory before we start sending the compressed bytes.

**Comment by nyh** [comment]
Pull request for this feature: https://github.com/scylladb/scylladb/pull/29669
